Add HttpVersion parameter to Invoke-WebRequest cmdlet

hayhay27 opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

Fix https://github.com/PowerShell/PowerShell/issues/12641

- Add `HttpVersionCompletionsAttribute` with completions `1.0`, `1.1`, `2.0`, `3.0`
- Handle input conversions from `string`, `int` and `double` to `Version` (generalize `ArgumentToVersionTransformationAttribute` from `Set-StrictMode`)
- Enrich Verbose mesasge for request with http version and uri
- Add pester tests for `HttpVersion` completions
- Add xUnit tests for `ArgumentToVersionTransformationAttribute`

## PR Context

This PR addressed to issue https://github.com/PowerShell/PowerShell/issues/12641
At the moment impossible make requests to network resources served only with HTTP/2 cause of default http version in `HttpRequestMessage.Version`=1.1

Changes in this PR allow to set http version during invocation `Invoke-RestMethod` or `Invoke-WebRequest`

This is my first PR to PowerShell. I have a couple of questions about tests. How and where to implement (at least point me right place please).

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed: https://github.com/MicrosoftDocs/PowerShell-Docs/issues/8065 <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed: <!-- Number/link of that issue here -->

hayhay27 committed:
Implement HttpVersion parameter for Invoke-RestMethod and Invoke-WebRequest

Added HttpVersionCompleter - 1.0, 1.1, 2.0, 3.0
Added HttpVersionTransformation supporting conversions from string, int and double
Enriched Verbose mesasge for request with http version and uri

hayhay27 committed:
Add HttpVersion parameter to Invoke-WebRequest

Fix code issues

hayhay27 committed:
Add HttpVersion parameter to Invoke-WebRequest

Fix code issue

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/HttpVersionCompleter.cs:
```diff
@@ -0,0 +1,32 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Collections;
```
We need this in every new file.
```suggestion

#nullable enable

using System.Collections;
```

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/HttpVersionCompleter.cs:
```diff
@@ -0,0 +1,32 @@
+namespace Microsoft.PowerShell.Commands
+{
+    /// <summary>
+    /// HttpVersionCompleter for http version names.
```
```suggestion
    /// A completer for HTTP version names.
```

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/HttpVersionTransformationAttribute.cs:
```diff
@@ -0,0 +1,36 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
```
```suggestion

#nullable enable

using System;
```

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/HttpVersionUtils.cs:
```diff
@@ -0,0 +1,42 @@
+
+        static HttpVersionUtils()
+        {
+            var fields = typeof(HttpVersion).GetFields(System.Reflection.BindingFlags.Static | System.Reflection.BindingFlags.Public);
```
Looks great but it is seem too expensive to use a reflection for so simple scenario. I'd prefer a static string array with well-know values.

hayhay27 replied:
There is very small impact. Static ctor called only once (when tab pressed first time). On the other hand we've got dynamic completion list. If it still doesn't fit i'll fix it.

iSazonov replied:
I have no strict objections and leave this to other reviewers.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/resources/WebCmdletStrings.resx:
```diff
@@ -1,17 +1,17 @@
-  <!--
-    Microsoft ResX Schema
-
+  <!-- 
+    Microsoft ResX Schema 
+    
```
Please revert extra spaces. Below too.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/HttpVersionTransformationAttribute.cs:
```diff
@@ -0,0 +1,36 @@
+                return inputData;
+            }
+
+            if (version is double versionDouble)
+            {
+                // 1.0 converts to string "1" so we preserve .0 with "0.0" format
+                return versionDouble.ToString("0.0", CultureInfo.InvariantCulture);
+            }
```
Does this work for "-Version 2"? I guess we get the argument as int type.

hayhay27 replied:
Do you mean `-HttpVersion 2`?
2 as int (handled bellow with other cases)
2.0 as double

iSazonov replied:
What if we implement a type converter so that LanguagePrimitives.ConvertTo() works for all needed cases? It seems then we can remove the transformation attribute and gets more benefits for users. 
Another thoughts we could implement the type converter and change the transformation attribute to process only well-known values like `Http10`, `Http20`, `Http11`, `Unknown `. It is for discussion.

I'd prefer the type converter and removing the attribute.

iSazonov replied:
Since we already have ArgumentToVersionTransformationAttribute I suggest to re-use it.

Thinking more about type converter I believe It can have undesirable side effects.

iSazonov reviewed (commented):
As for test, I think we have no need to do real HTTP requests with specified version since we utilize .Net Runtime features. I think it would enough if you did manual checks and shared results.
Nevertheless, we must verify that (1) the parameter accepts well-know values without throw, (2) transformation attribute works (it could be xUnit tests), (3) tab-completion works.

hayhay27 committed:
Apply suggestions from code review

Add nullable directive
Fix comment

Co-authored-by: Ilya <darpa@yandex.ru>

hayhay27 committed:
Add HttpVersion parameter to Invoke-WebRequest

Revert unrelated changes

hayhay27 committed:
Add HttpVersion parameter to Invoke-WebRequest

Add tests for version completer
Add tests for version transformation

hayhay27 commented:
> 
> 
> As for test, I think we have no need to do real HTTP requests with specified version since we utilize .Net Runtime features. I think it would enough if you did manual checks and shared results.
> Nevertheless, we must verify that (1) the parameter accepts well-know values without throw, (2) transformation attribute works (it could be xUnit tests), (3) tab-completion works.

1. do you mean a pester test?
2. done with xunit
3. done with xunit

hayhay27 committed:
Add HttpVersion parameter to Invoke-WebRequest

fix code issues

hayhay27 committed:
Add HttpVersion parameter to Invoke-WebRequest

fix code issues

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/HttpVersionUtils.cs:
```diff
@@ -0,0 +1,42 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
```
```suggestion
// Licensed under the MIT License.

#nullable enable

using System;
```

hayhay27 commented on test/xUnit/csharp/test_HttpVersionTransformationAttribute.cs:
```diff
@@ -0,0 +1,76 @@
+            Assert.Equal(expected, result);
+        }
+
+        public static IEnumerable<object[]> IntCases()
+        {
+            yield return new object[] { 1, new Version(1, 0) };
+            yield return new object[] { 2, new Version(2, 0) };
```
Make sense.
I took `ArgumentToVersionTransformationAttribute` as a basis. As a common approach.

iSazonov replied:
In the case it makes sense to re-use  ArgumentToVersionTransformationAttribute.

hayhay27 replied:
Don't think so. That attribute specific for SetStrictCommand and useless without customization.

iSazonov replied:
We should Dry principle.

My proposal:
1.  Introduce `VersionTransformationAttribute` as common ancestor to enhance standard type converter 
2.  In ArgumentToVersionTransformationAttribute and HttpVersionTransformationAttribute at first place process string values then fallback to VersionTransformationAttribute 
3. HttpVersionCompleter return string values. This makes scripts more readable.

hayhay27 replied:
1 and 2: You're right about transformation.
3: As I understood we must return quoted values on completion.

iSazonov replied:
> As I understood we must return quoted values on completion.

No. Quotes are needed only if there are spaces in names.

hayhay27 commented on test/xUnit/csharp/test_HttpVersionTransformationAttribute.cs:
```diff
@@ -0,0 +1,76 @@
+            Assert.Equal(expected, result);
+        }
+
+        public static IEnumerable<object[]> PSObjectCases()
+        {
+            yield return new object[] { new PSObject(false), new Version(0, 0) };
+            yield return new object[] { new PSObject(true), new Version(1, 0) };
```
I assumed this is the common approach to handle PSObject in transformations with fallback to `LanguagePrimitives.TryConvertTo`
Again like in `ArgumentToVersionTransformationAttribute` )

hayhay27 committed:
Add HttpVersion parameter to Invoke-WebRequest

generalize ArgumentToVersionTransformationAttribute
move xUnit tests to Pester

hayhay27 committed:
Add HttpVersion parameter to Invoke-WebRequest

fix documentation format

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/resources/WebCmdletStrings.resx:
```diff
@@ -129,7 +129,7 @@
     <value>The cmdlet cannot run because the following parameter is not specified: Credential. The supplied Authentication type requires a Credential. Specify Credential, then retry.</value>
   </data>
-    <data name="AuthenticationTokenNotSupplied" xml:space="preserve">
+  <data name="AuthenticationTokenNotSupplied" xml:space="preserve">
```
Indentation. Please revert.

hayhay27 replied:
Why? Somebody add two unnecessary spaces.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/HttpVersionCompleter.cs:
```diff
@@ -0,0 +1,39 @@
+
+            var wordToCompletePattern = WildcardPattern.Get(string.IsNullOrWhiteSpace(valueToComplete) ? "*" : valueToComplete + "*", WildcardOptions.IgnoreCase);
+
+            foreach (var version in HttpVersionUtils.AllowedVersions)
+            {
+                if (wordToCompletePattern.IsMatch(version))
+                {
+                    var quotedVersion = Quoted(version);
+                    yield return new CompletionResult(quotedVersion, quotedVersion, CompletionResultType.ParameterValue, quotedVersion);
+                }
+            }
```
I think it is more readable to show version names to users, not values.
So please replace `var` with `string` and remove Quoted() since no spaces are in the names.

hayhay27 replied:
1.0 treated as double without quoting
Do you still thinking that we must instantiate Version in all cases during transformation? I wouldn't to change current behavior for Set-StrictMode

iSazonov replied:
> 1.0 treated as double without quoting

If we want to use string names (like HTTP3) it is not problem. Direct using double is also not problem.

> Do you still thinking that we must instantiate Version in all cases during transformation? I wouldn't to change current behavior for Set-StrictMode

Current implementation has implicit behaviors in transformation attribute. I believe we could do the work explicitly and this will allow us to use dry principle for new (HttpVersion) attribute.

hayhay27 replied:
Ilya, I'm confused )
Are you ok with completion 1.0, 1.1, 2.0, 3.0? Or you want to change it to HTTP1, HTTP1.1, etc?
Are you ok with my current implementation of ArgumentToVersionTranformationAttribute? I made class extensible

iSazonov replied:
We usually discuss this in issue before PR. But since it is still draft we can consider alternatives.
I think `-HttpVersion HTTP21` looks more readable than `-HttpVersion 2.1`. This follows "Latest" in Set-Strict. If you prefer second one (double) I have no objections since we can add first one later without a breaking change. Also we can refactor existing code later (Ideally we could have one universal parametrized attribute for both cases.)

So you have two options to choose from:
- simply add enhancement as you started
- continue to implement common solution with code refactoring.

hayhay27 replied:
As I understood, you propose to do something like this (to be universal enough):

```csharp
public TransformationAttribute
{

    private readonly IReadOnlyDictionary<string, Version> _map;

    public TransformationAttribute(Type mappingProvider)
    {
        _map = GetValues(mappingProvider);
    }

    // other code
}

// Invoke-WebRequest
[TransformationAttribute(typeof(HttpVersionMapping))]
// Set-StrictMode
[TransformationAttribute(typeof(SetStrictModeVersionMapping))]
```

According to [specification](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/attributes#attribute-parameter-types) we cannot create universal parametrized implementation other than via `Type` argument with implicit convention.

And I think this not worth to do it. At least for now.

hayhay27 replied:
I hope this is it )

hayhay27 committed:
Add HttpVersion parameter to Invoke-WebRequest

generalize version transformation attribute

hayhay27 committed:
Add HttpVersion parameter to Invoke-WebRequest

fix code style
fix pester tests

iSazonov commented on src/System.Management.Automation/engine/ArgumentToVersionTransformationAttribute.cs:
```diff
@@ -0,0 +1,77 @@
+    ///   * A simple integer, i.e. 2;
+    ///   * A string without a dot, i.e. "2".
+    /// </summary>
+    public class ArgumentToVersionTransformationAttribute : ArgumentTransformationAttribute
```
private/internal sealed?

hayhay27 replied:
make just as internal

`sealed` blocks inheritance and `private` doesn't cover by `InternalsVisibleTo` attribute

iSazonov commented on test/xUnit/csharp/test_ArgumentToVersionTransformationAttribute.cs:
```diff
@@ -0,0 +1,46 @@
+            Assert.Equal(expected, result);
+        }
+
+        public static IEnumerable<object[]> TestCases()
```
I'd add negative tests too.

hayhay27 replied:
update `unhandled` section

hayhay27 committed:
Add HttpVersion parameter to Invoke-WebRequest

add tests cases
make ArgumentToVersionTransformationAttribute internal

hayhay27 commented:
I have no idea about failed tests )
What is that? )

iSazonov committed:
Update src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/HttpVersionUtils.cs

hayhay27 committed:
Add HttpVersion parameter to Invoke-WebRequest

add NRE handling

hayhay27 reviewed (commented):
> Please update the PR description to address latest code changes.

done

hayhay27 commented:
while we keep silence

I thought with your blessing I can update docs for this PR

iSazonov commented:
You can open new issue in PowerShell-Docs repository and add reference in the PR description.

daxian-dbw commented:
@hayhay27 Thanks you for your contribution!
A new parameter `-HttpVersion` is added but there is no test case to validate the use of this new parameter. Can you please add functional tests to validate if this new parameter works as expected?

iSazonov commented:
@daxian-dbw Can you confirm that CI OS-s we use support HTTP3 and it is enabled by default? I think we can do only manual check today.

hayhay27 commented:
> @daxian-dbw Can you confirm that CI OS-s we use support HTTP3 and it is enabled by default? I think we can do only manual check today.

We can do it only for 1.0, 1.1, 2.0. I think we can extrapolate results to 3.0
Could you point me same tests (right place) where I can add new one.

daxian-dbw commented:
I don't know whether the VMs used in CI support HTTP3, but yes, at least validate the new parameter for 1.0, 1.1, and 2.0.
You can find the `WebCmdlets.test.ps1` at https://github.com/PowerShell/PowerShell/blob/master/test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1, which contains all tests for `Invoke-WebRequest` and `Invoke-RestMethod`.

I already requested the Cmdlet WG to review this PR and the corresponding issue.

hayhay27 committed:
Add HttpVersion parameter to Invoke-WebRequest

add pester tests

hayhay27 committed:
Add HttpVersion parameter to Invoke-WebRequest

skip tests for HttpVersion on mac and linux

iSazonov commented:
@hayhay27 Please look CodeFactor issues.

hayhay27 commented:
> 
> 
> @hayhay27 Please look CodeFactor issues.

Just one question. How to check code style locally?

iSazonov commented:
> > @hayhay27 Please look CodeFactor issues.
> 
> Just one question. How to check code style locally?

You have no need to do this locally - see CofeFactor report in the PR.
(but you could temporary add the package https://github.com/DotNetAnalyzers/StyleCopAnalyzers)

hayhay27 committed:
Add HttpVersion parameter to Invoke-WebRequest

fix code smells

hayhay27 commented:
@iSazonov
Hi! What can I do to complete this PR?

daxian-dbw commented:
@hayhay27 This PR and the corresponding issue need to be reviewed by Cmdlet Working Group. I just pinged that WG again. Thank you for your patience!

rjmholt commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/HttpVersionUtils.cs:
```diff
@@ -0,0 +1,47 @@
+
+namespace Microsoft.PowerShell.Commands
+{
+    internal static class HttpVersionUtils
```
Given the fact that this is only used on the attribute, the attribute is internal and is only used in a specific context, it feels to me like the functionality in this class just belongs in the attribute class instead

rjmholt commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -2076,6 +2092,20 @@ Describe "Invoke-RestMethod tests" -Tags "Feature", "RequireAdminOnWindows" {
 
+    $httpVersions = @('1.1', '2')
+    foreach ($httpVersion in $httpVersions) {
+        It "Validate Invoke-RestMethod -HttpVersion $httpVersion" -Skip:(!$IsWindows) {
```
This should probably be a `-TestCases` use rather than inside a `foreach` loop -- especially since that will make migrating to Pester 5 easier

rjmholt commented:
The changes in this PR look good and I think the actual feature addition is great. Made some style comments

hayhay27 committed:
Add HttpVersion parameter to Invoke-WebRequest

fixes by review
- remove HttpVersionUtils
- refactor pester test

hayhay27 commented:
@rjmholt Thanks a lot!
I applied all your comments

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/HttpVersionCompletionsAttribute.cs:
```diff
@@ -0,0 +1,57 @@
+
+        static HttpVersionCompletionsAttribute()
+        {
+            FieldInfo[] fields = typeof(HttpVersion).GetFields(System.Reflection.BindingFlags.Static | System.Reflection.BindingFlags.Public);
```
```suggestion
            FieldInfo[] fields = typeof(HttpVersion).GetFields(BindingFlags.Static | BindingFlags.Public);
```

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/HttpVersionCompletionsAttribute.cs:
```diff
@@ -0,0 +1,57 @@
+        {
+            FieldInfo[] fields = typeof(HttpVersion).GetFields(System.Reflection.BindingFlags.Static | System.Reflection.BindingFlags.Public);
+
+            if (fields.Length == 0)
+            {
+                AllowedVersions = Array.Empty<string>();
+                return;
+            }
```
This seems unneeded, as it will be breaking change if `HttpVersion` removes those fields in future.

daxian-dbw commented on src/System.Management.Automation/engine/ArgumentToVersionTransformationAttribute.cs:
```diff
@@ -0,0 +1,77 @@
+        /// Initializes a new instance of the ArgumentToVersionTransformationAttribute class.
+        /// </summary>
+        /// <param name="map">Version mapping.</param>
+        public ArgumentToVersionTransformationAttribute(IReadOnlyDictionary<string, Version> map)
+        {
+            _map = map ?? _empty;
+        }
```
I don't think this constructor is useful to an attribute type, because you cannot specify a non-constant value when declaring an attribute to a parameter. This should be removed.

daxian-dbw commented on src/System.Management.Automation/engine/ArgumentToVersionTransformationAttribute.cs:
```diff
@@ -0,0 +1,77 @@
+    /// </summary>
+    internal class ArgumentToVersionTransformationAttribute : ArgumentTransformationAttribute
+    {
+        private static readonly IReadOnlyDictionary<string, Version> _empty = new Dictionary<string, Version>();
+
+        private readonly IReadOnlyDictionary<string, Version> _map;
```
By looking at the code, the sole purpose of having this dictionary is for `latest` to work for `Set-StrictMode` command.
Let's not depending on this dictionary. Instead, add a new virtual method `virtual Version ConvertFromString(string )`, which simply returns `null` in this type. Then the transformation attribute for `Set-StrictMode` can handle `latest` in its override method.

When `ConvertFromString` returns `null`, proceed on with the other transformation steps, otherwise, return the result directly.

daxian-dbw commented on src/System.Management.Automation/engine/InternalCommands.cs:
```diff
@@ -2647,46 +2647,17 @@ public SwitchParameter Off
+        /// Handle 'latest', which we interpret to be the current version of PowerShell.
         /// </summary>
-        private sealed class ArgumentToVersionTransformationAttribute : ArgumentTransformationAttribute
+        private sealed class CommandArgumentToVersionTransformationAttribute : ArgumentToVersionTransformationAttribute
```
```suggestion
        private sealed class ArgumentToPSVersionTransformationAttribute : ArgumentToVersionTransformationAttribute
```

iSazonov replied:
We convert to Version, not to PSVersion.

daxian-dbw replied:
The use of this particular attribute is for `Set-StrictMode -Version`, and the argument must be valid PowerShell version.
```
> Set-StrictMode -Version 3.5
Set-StrictMode: Cannot validate argument on parameter 'Version'. The "3.5" argument does not contain a valid PowerShell version. Supply a valid version number and then try the command again.
```

hayhay27 committed:
Add HttpVersion parameter to Invoke-WebRequest

fixes by review
- extend version transformation via method override instead of ctor injection
- clean up code

daxian-dbw reviewed (approved):
LGTM

daxian-dbw commented:
@hayhay27 Thank you for your contribution!
